### PR TITLE
GDLauncher v2

### PR DESCRIPTION
**Pull request by blarfoon**
hi

### CODE REVIEW

**Review by fulviodenza — APPROVED** [review]

**Review by fulviodenza — APPROVED** [review]
LGTM

File: README.md [hunk 1]
```diff
@@ -101,7 +101,7 @@ These are the steps to compile it yourself.
 
 You need the following software installed:
 
-- [NodeJS](https://nodejs.org/en/download/) (> v14.14.0 x64)
+- [NodeJS](https://nodejs.org/en/download/) (> v16.13.1 x64)
 - [Rust](https://www.rust-lang.org/)
 - [Python2](https://www.python.org/)(Mac and linux)
 - C++ compiler (g++ or windows build tools)
```
Comment (Eskaan) [inline README.md]: Add Go 17 here.
Also what things here do we even need anymore?

File: gdlib/FSWatcher.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	for _, v := range dirsToWatch {
+		if v == directory {
+			alreadyWatching = true
+			break
+		}
+	}
+	if alreadyWatching {
+		done <- errors.New("directory already being watched")
+		return
+	}
```
Comment (Eskaan) [inline gdlib/FSWatcher.go]: Just move it into the loop.
(You said you fixed it already, just resolve when deploying)

**Comment by Kaliendo** [comment]
LGTM

**Review by Kaliendo — APPROVED** [review]
LGTM!

File: gdlib/FSWatcher.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		select {
+		case <-stopChans[directory]:
+			delete(stopChans, directory)
+			return
+		case <-time.After(200 * time.Millisecond):
+			{
+				filesOnDisk, err := readAllFiles(directory)
+				if err != nil {
+					log.Fatal(err) // TODO: FIX
+				}
```
Comment (Eskaan) [inline gdlib/FSWatcher.go]: What exactly needs to be fixed here? Does it fail when there is no read permission on the directory?

File: gdlib/adapters/socket.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+	FS_WATCHER_LIST
+)
+
+var shouldQuit = true
+var upgrader = websocket.Upgrader{}
+var semaphore = make(chan int, 1)
+
+var quitError = make(chan error)
+
+const PORT = 7890
```
Comment (Eskaan) [inline gdlib/adapters/socket.go]: Will this port be running open to the local network in the future? If yes, we should see that we only accept requests from localhost

File: gdlib/adapters/socket.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+func StartServer() error {
+	// If we don't receive a ping signal within 10 seconds, we should quit
+	go func() {
+		// Intentionally not passing shouldQuit as parameter because we want to
+		// always read the latest value. This should usually be considered a data
+		// race but it isn't in this case
+		time.Sleep(10 * time.Second)
+		if shouldQuit {
+			quitError <- errors.New("quitting caused by no ping received within 10s from startup")
```
Comment (Eskaan) [inline gdlib/adapters/socket.go]: Little typo, just capitalise and set a dot.

File: gdlib/adapters/socket.go (hunk 4 above)
Comment (Eskaan) [reply gdlib/adapters/socket.go]: nvm I think you are doing that in line 98, resolve it when I am right

File: gdlib/adapters/socket.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+	actionParam, ok := payload["action"]
+	if !ok {
+		return 1, errors.New("path not specified in payload")
+	}
+
+	action, ok := actionParam.(float64)
+	if !ok {
+		return 1, errors.New("path not a float64")
+	}
```
Comment (Eskaan) [inline gdlib/adapters/socket.go]: Error says its the path, change that to action

File: gdlib/adapters/socket.go [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+	var done = make(chan error)
+	switch action {
+	default:
+	case FS_WATCHER_START:
+		go gdlib.StartFSWatcher(directory, updateFunc, done)
+	case FS_WATCHER_STOP:
+		go gdlib.StopFSWatcher(directory, updateFunc, done)
+	case FS_WATCHER_LIST:
+		fmt.Println("FSWatcher: List")
```
Comment (Eskaan) [inline gdlib/adapters/socket.go]: // Todo add listing feature?

**Review by Eskaan — CHANGES_REQUESTED** [review]

**Review by fulviodenza — APPROVED** [review]

**Pull request closed without merging**